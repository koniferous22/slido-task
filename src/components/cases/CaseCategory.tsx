import React from 'react'

import { Case } from 'types/common/Case'
import { CaseCategoryProps } from 'types/cases/CaseCategory'

import { shortenTitle } from 'utils/functions'

import 'styles/cases/CaseCategory.css'

const CaseCategory: React.FC<CaseCategoryProps> = ({id, label, caseEntries, offset, handleSelectCase}: CaseCategoryProps) => {
	const listEntries = caseEntries.map((case_: Case, index: number) => (
			<li 
				key={index} 
				onClick={(): void => handleSelectCase(index + (offset || 0))}
			>
				<i>{'[' + case_.date.toLocaleDateString() + '] '}</i>
				{shortenTitle(case_.title, 50)}
			</li>		
		))
	return (
		<section id={id}>
			<h4 className="case-list-title">{label}</h4>
			<ul className="case-list">
			{
				listEntries
			}
			</ul>
		</section>
	)	
}

export default CaseCategory
