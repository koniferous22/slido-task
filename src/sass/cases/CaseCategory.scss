@use '../_base' as base;

%case-list-section {
	max-width: base.$case-list-max-width;
}

section#cases {
	div#cases-hint {
		flex-shrink: 2;
		@include base.disable-small-devices(600px);
	}
	 

	section#past {
		@extend %case-list-section;
		transform: base.$cool-transformation-left;
	}


	section#future {
		@extend %case-list-section;
		transform: base.$cool-transformation-right;
	}

	ul.case-list {
		@include base.scrollable-case-list;

		transition: transform base.$transition-time;
		list-style-type: none;
		
		/*disable scrollbar in firefox*/
		scrollbar-width: none;
	}

	@include base.case-list-scrollbar;

	ul.case-list:after {
		@include base.bottom-border-overlay(245, 70%);
	}

	ul.case-list li {
		transition: transform base.$transition-time;
	}

	ul.case-list li:hover {
		transform: scale(1.1);
		cursor: pointer;
	}
}

