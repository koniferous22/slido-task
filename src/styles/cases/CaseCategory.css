section#cases section {
	max-width: 30vw;
	overflow-x: auto;
	flex-grow: 30000;
    text-align: center;
}

section#cases section#past {
	transform: matrix3d(
    	1, 0, 0, 0.001,
    	0.1, 1, 0, 0,
    	0, 0, 1, 0,
    	0, 0, 0, 1);
}

section#cases section#future {
	transform: matrix3d(
    	1, 0, 0, -0.001,
    	-0.1, 1, 0, 0,
    	0, 0, 1, 0,
    	0, 0, 0, 1);
	/*margin-right: 50px;*/
}

section#cases ul.case-list {
    max-height: 20vh;
	overflow-y: scroll;
    padding: 10px 5vw 50px 5vw;
    border-radius: 20px;

    transition: transform .2s;
    /*disable scrollbar in firefox*/
    scrollbar-width: none;

    list-style-type: none;
}

section#cases ul.case-list:hover {
	scrollbar-width: thin;
}

/*disable scrollbar in webkit based browsers*/
section#cases ul.case-list::-webkit-scrollbar {
   display: none;
}


section#cases ul.case-list:after {
    content: "";
    display: block;
    width: 100%;
    height: 100%;
    background: linear-gradient(to bottom, rgba(245,245,245,0) 0%, rgba(245,245,245,0) 70%, rgba(245,245,245,1) 100%);
    pointer-events: none;
    position: absolute;
    top: 0;
    left: 0;
}

section#cases ul.case-list li {
	transition: transform .2s;
}

section#cases ul.case-list li:hover {
	transform: scale(1.1);
	cursor: pointer;
}

