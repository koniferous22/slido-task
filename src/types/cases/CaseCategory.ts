import { Case } from 'types/common/Case'

export interface CaseCategoryProps {
	id: string;
	label: string;
	caseEntries: Array<Case>;
	handleSelectCase: (_: number) => void;
	offset?: number;
}
