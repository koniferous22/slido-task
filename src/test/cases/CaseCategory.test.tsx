import React from 'react'
import renderer from 'react-test-renderer'
import CaseCategory from 'components/cases/CaseCategory'

import { cleanup, fireEvent, render } from '@testing-library/react'

import { shortenTitle } from 'utils/functions'

import { sampleCases } from 'test/data/sampleCases'

afterEach(cleanup)

it('CaseCategory renders correctly', () => {	
	const tree = renderer
		.create(
			<CaseCategory
				id="case-list"
				label="Case List"
				caseEntries={sampleCases}
				handleSelectCase={(): void => {return}}
			/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});


it('Select case event', () => {
	const mockSelectCase = jest.fn((): void => {return});
	const { getByText, findByText } = render(<CaseCategory
		id="case-list"
		label="Case List"
		caseEntries={sampleCases}
		handleSelectCase={mockSelectCase}
	/>)
	fireEvent.click(getByText(shortenTitle(sampleCases[0].title, 50)));
	expect(mockSelectCase).toHaveBeenCalled()
})