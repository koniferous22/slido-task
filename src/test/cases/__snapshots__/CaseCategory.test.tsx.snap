// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseCategory renders correctly 1`] = `
<section
  id="case-list"
>
  <h4
    className="case-list-title"
  >
    Case List
  </h4>
  <ul
    className="case-list"
  >
    <li
      onClick={[Function]}
    >
      <i>
        [1/26/2020] 
      </i>
      Dni Zel√° v Stupave
    </li>
    <li
      onClick={[Function]}
    >
      <i>
        [2/26/2020] 
      </i>
      Filakovsky jarmok
    </li>
    <li
      onClick={[Function]}
    >
      <i>
        [2/26/2020] 
      </i>
      I'm Rick Harrison, and this is my pawn shop.I work...
    </li>
  </ul>
</section>
`;
